Ensure that compiler output is sufficiently aligned.

stellaraccident opened the pull request:
On modern Linux, memfd_create is used, which is always page aligned. However on others (Windows/MacOS), we fall back to a malloc-based implementation, which is not.

Adds aligned allocation fallback so we always meet this invariant.

Fixes #14926

kuhar commented on compiler/src/iree/compiler/API/Internal/CompilerDriver.cpp:
```diff
@@ -90,6 +89,74 @@
+template <typename T>
+struct rt_aligned_allocator {
+  using value_type = T;
+  rt_aligned_allocator() noexcept {}
+  template <class U>
+  rt_aligned_allocator(const rt_aligned_allocator<U> &) noexcept {}
```
Do we expect IREE to be ever built with exceptions enabled and benefit from `noexcept`?

stellaraccident replied:
No. But allocators are finicky and this is the exact standard compliant way to define one.

kuhar commented on compiler/src/iree/compiler/API/Internal/CompilerDriver.cpp:
```diff
@@ -90,6 +89,74 @@
 namespace mlir::iree_compiler::embed {
 namespace {
 
+// If note using memfd_create, then we need to align output buffers
```
nit: did you mean 'not'?
```suggestion
// If not using memfd_create, then we need to align output buffers
```

stellaraccident committed:
Update compiler/src/iree/compiler/API/Internal/CompilerDriver.cpp

kuhar reviewed (approved):
LGTM

stellaraccident committed:
Ensure that compiler output is sufficiently aligned.

On modern Linux, memfd_create is used, which is always page aligned. However on others (Windows/MacOS), we fall back to a malloc-based implementation, which is not.

Adds aligned allocation fallback so we always meet this invariant.

Fixes #14926

stellaraccident committed:
Typo

antiagainst commented on compiler/src/iree/compiler/API/Internal/CompilerDriver.cpp:
```diff
@@ -516,6 +586,8 @@ Error *Output::openMembuffer() {
   // for memfiles.
   type = Type::Membuffer;
+  // Avoid some initial memcpys with a size appropriate for program output.
+  outputString.reserve(16384);
```
Nit: might want to define a constant at the beginning of the file for this magic number?

stellaraccident replied:
I'll do so in a follow-up. There's another place in the compiler I need this, so it will be moving to a utility header soon.
